### PR TITLE
common: sort best-matched commond by req argument count

### PR DESCRIPTION
**Pull request by liuchang0812**
Fixes: http://tracker.ceph.com/issues/40292

Signed-off-by: Chang Liu <liuchang0812@gmail.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by liuchang0812** [comment]
@tchaikov mind taking a look? thanks

File: src/pybind/ceph_argparse.py [hunk 1]
```diff
@@ -1154,9 +1154,10 @@ def validate_command(sigdict, args, verbose=False):
         else:
             bestcmds.append(cmd)
 
-    # Sort bestcmds by number of args so we can try shortest first
+    # Sort bestcmds by number of req args so we can try shortest first
     # (relies on a cmdsig being key,val where val is a list of len 1)
-    bestcmds_sorted = sorted(bestcmds, key=lambda c: len(c['sig']))
+    is_req_arg = lambda arg: not hasattr(arg, "req") or arg.req == True
+    bestcmds_sorted = sorted(bestcmds, key=lambda c: len([i for i in c['sig'] if is_req_arg(i)]))
```
Comment (tchaikov) [inline src/pybind/ceph_argparse.py]: i think `hasattr(arg, "req")` is always true. and i'd suggest something like:

```py
def grade(cmd):
  # prefer optional arguments over required ones
  sigs = cmd['sig']
  return sum(map(lambda sig: [0.5, 1][sig.req], sigs))
bestcmds_sorted = sorted(bestcmds, key=grade)
```

Comment (liuchang0812) [reply src/pybind/ceph_argparse.py]: thank you. your patch can't pass following case, see

```
./bin/ceph progress --verbose
bestcmds_sorted: 
[{u'flags': 8,
  u'help': u'Reset progress tracking',
  u'module': u'mgr',
  u'perm': u'rw',
  u'sig': [argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=progress),
           argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=clear)]},
 {u'flags': 8,
  u'help': u'Show progress of recovery operations',
  u'module': u'mgr',
  u'perm': u'r',
  u'sig': [argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=progress),
           argdesc(<class 'ceph_argparse.CephInt'>, name=event_id, numseen=0, req=False, n=1),
           argdesc(<class 'ceph_argparse.CephInt'>, name=test, numseen=0, req=False, n=1)]},
 {u'flags': 8,
  u'help': u'Show machine readable progress information',
  u'module': u'mgr',
  u'perm': u'r',
  u'sig': [argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=progress),
           argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=json),
           argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=test)]}]
no valid command found; 3 closest matches:
progress clear
progress {<int>} {<int>}
progress json test
Error EINVAL: invalid command
```

Comment (tchaikov) [reply src/pybind/ceph_argparse.py]: we can make it simpler then
```py
def grade(cmd):
  # prefer optional arguments over required ones
  sigs = cmd['sig']
  return sum(map(lambda sig: sig.req, sigs))
bestcmds_sorted = sorted(bestcmds, key=grade)
```

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-06-24_01:13:41-rados-wip-kefu-testing-2019-06-23-2322-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-06-25_00:46:45-rados-wip-kefu-testing-2019-06-23-2322-distro-basic-smithi/

the failure are tracked by http://tracker.ceph.com/issues/40533

**Pull request merged**